Windows 10 build errors

This issue was created by my mistake on ULS/STIR (issue 461), reproducing here now.

Building SIRF on my Windows 10 laptop I get:
```
1>C:\Users\wps46139\Documents\GitHub\build\SIRF-SuperBuild\sources\SIRF\src\Registration\NiftyMoMo\BSplineTransformation.cpp(381): error C3028: 'NiftyMoMo::BSplineTransformation::controlPointGridImage': only a variable or static data member can be used in a data-sharing clause
1>C:\Users\wps46139\Documents\GitHub\build\SIRF-SuperBuild\sources\SIRF\src\Registration\NiftyMoMo\BSplineTransformation.cpp(527): error C3028: 'NiftyMoMo::BSplineTransformation::controlPointGridImage': only a variable or static data member can be used in a data-sharing clause
```
Commenting out the two offendng #pragma omp, I get another error caused apparently by `NiftiImageData<dataType>::get_inner_product()`:
```
7>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\numeric(98): error C2794: 'iterator_category': is not a member of any direct or indirect base class of 'std::iterator_traits<_InIt1>'
7>          with
7>          [
7>              _InIt1=sirf::NiftiImageData<float>::Iterator_const
7>          ]
7>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\numeric(111): note: see reference to function template instantiation '_Ty std::_Inner_product_no_deprecate<_InIt1,_InIt2,_Ty,_Fn21,_Fn22>(_InIt1,_InIt1,_InIt2,_Ty,_Fn21 &,_Fn22 &)' being compiled
7>          with
7>          [
7>              _Ty=float,
7>              _InIt1=sirf::NiftiImageData<float>::Iterator_const,
7>              _InIt2=sirf::NiftiImageData<float>::Iterator_const,
7>              _Fn21=std::plus<void>,
7>              _Fn22=std::multiplies<void>
7>          ]
7>  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\numeric(139): note: see reference to function template instantiation '_Ty std::inner_product<_InIt1,_InIt2,_Ty,std::plus<void>,std::multiplies<void>>(_InIt1,_InIt1,_InIt2,_Ty,_Fn21,_Fn22)' being compiled
7>          with
7>          [
7>              _Ty=float,
7>              _InIt1=sirf::NiftiImageData<float>::Iterator_const,
7>              _InIt2=sirf::NiftiImageData<float>::Iterator_const,
7>              _Fn21=std::plus<void>,
7>              _Fn22=std::multiplies<void>
7>          ]
7>  C:\Users\wps46139\Documents\GitHub\build\SIRF-SuperBuild\sources\SIRF\src\Registration\cReg\NiftiImageData.cpp(1438): note: see reference to function template instantiation '_Ty std::inner_product<sirf::NiftiImageData<dataType>::Iterator_const,sirf::NiftiImageData<dataType>::Iterator_const,dataType>(_InIt1,_InIt1,_InIt2,_Ty)' being compiled
7>          with
7>          [
7>              _Ty=float,
7>              dataType=float,
7>              _InIt1=sirf::NiftiImageData<float>::Iterator_const,
7>              _InIt2=sirf::NiftiImageData<float>::Iterator_const
7>          ]
7>  C:\Users\wps46139\Documents\GitHub\build\SIRF-SuperBuild\sources\SIRF\src\Registration\cReg\NiftiImageData.cpp(1437): note: while compiling class template member function 'float sirf::NiftiImageData<dataType>::get_inner_product(const sirf::NiftiImageData<dataType> &) const'
7>          with
7>          [
7>              dataType=float
7>          ]
7>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\numeric(98): error C2938: '_Iter_cat_t<sirf::NiftiImageData<dataType>::Iterator_const>' : Failed to specialize alias template
7>          with
7>          [
7>              dataType=float
7>          ]
7>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\numeric(98): error C2062: type 'unknown-type' unexpected
```
After commenting out all occurences of `get_inner_product()` in C++, Matlab and Python, I finally managed to build all projects except `SYN_TESTS_CPLUSPLUS`, which fails with this weird message:
```
4>C:\Users\wps46139\Documents\GitHub\build\SIRF-SuperBuild\INSTALL\include\boost-1_65_1\boost/asio/detail/socket_types.hpp(24): fatal error C1189: #error:  WinSock.h has already been included
```
(even more weirdly, my Visual Studio solution does not even have this project in the Solution Explorer window - its name only appears in the Error List window). After that, at least all non-MR tests run ok except `SYN_TESTS_CPLUSPLUS`, which is skipped by `ctest`, and `SYN_TEST_PYTHON`, which throws exception after about 7 sec.

--- a/src/iUtilities/include/sirf/iUtilities/DataHandle.h
+++ b/src/iUtilities/include/sirf/iUtilities/DataHandle.h
@@ -342,8 +342,11 @@ inline void* charDataHandleFromCharData(const char* s)
 	DataHandle* h = new DataHandle;
 	size_t len = strlen(s);
 	char* d = (char*)malloc(len + 1);
-	//strcpy_s(d, len + 1, s);
+#ifdef _MSC_VER
+	strcpy_s(d, len + 1, s);
+#else
 	strcpy(d, s);
+#endif
 	h->set((void*)d, 0, GRAB);
 	return (void*)h;
 }

--- a/src/xGadgetron/cGadgetron/gadgetron_x.cpp
+++ b/src/xGadgetron/cGadgetron/gadgetron_x.cpp
@@ -27,14 +27,28 @@ limitations under the License.
 \author Evgueni Ovtchinnikov
 \author SyneRBI
 */
-
 #include "sirf/iUtilities/DataHandle.h"
 #include "sirf/Gadgetron/cgadgetron_shared_ptr.h"
 #include "sirf/Gadgetron/gadgetron_data_containers.h"
 #include "sirf/Gadgetron/gadgetron_x.h"
+#include "sirf/Gadgetron/gadgetron_client.h"
 
 using namespace gadgetron;
 using namespace sirf;
+
+GTConnector::GTConnector()
+{
+	sptr_con_ = gadgetron::shared_ptr < GadgetronClientConnector >
+		(new GadgetronClientConnector);
+}
+GadgetronClientConnector& GTConnector::operator()()
+{
+	return *sptr_con_.get();
+}
+gadgetron::shared_ptr<GadgetronClientConnector> GTConnector::sptr()
+{
+	return sptr_con_;
+}
 
 static bool
 connection_failed(int nt)

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_client.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_client.h
@@ -63,6 +63,10 @@ IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #ifndef GADGETRON_CLIENT
 #define GADGETRON_CLIENT
 
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+
 #include <chrono>
 #include <condition_variable>
 #include <exception>

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h
@@ -40,8 +40,6 @@ limitations under the License.
 
 #include <ismrmrd/ismrmrd.h>
 #include <ismrmrd/dataset.h>
-#include <ismrmrd/meta.h>
-#include <ismrmrd/xml.h>
 
 #include "sirf/common/DataContainer.h"
 #include "sirf/common/MRImageData.h"

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_data_containers.h
@@ -289,8 +289,8 @@ namespace sirf {
 
 		int index(int i) const
 		{
-			int ni = index_.size();
-			if (ni > 0 && i >= ni || i < 0 || i >= number())
+			const std::size_t ni = index_.size();
+			if (i < 0 || (ni > 0 && static_cast<std::size_t>(i) >= ni) || static_cast<unsigned>(i) >= number())
 				THROW("Aquisition number is out of range");
 			if (ni > 0)
 				return index_[i];
@@ -594,8 +594,8 @@ namespace sirf {
 		const std::vector<int>& index() const { return index_; }
 		int index(int i) const
 		{
-			int ni = index_.size();
-			if (ni > 0 && i >= ni || i < 0 || i >= number())
+			const std::size_t ni = index_.size();
+			if (i < 0 || (ni > 0 && static_cast<std::size_t>(i) >= ni) || static_cast<unsigned>(i) >= number())
 				THROW("Image number is out of range");
 			if (ni > 0)
 				return index_[i];

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_image_wrap.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_image_wrap.h
@@ -1,7 +1,7 @@
 /*
 SyneRBI Synergistic Image Reconstruction Framework (SIRF)
 Copyright 2015 - 2020 Rutherford Appleton Laboratory STFC
-Copyright 2020 University College London
+Copyright 2020, 2021 University College London
 
 This is software developed for the Collaborative Computational
 Project in Synergistic Reconstruction for Biomedical Imaging (formerly CCP PETMR)
@@ -83,6 +83,12 @@ limitations under the License.
 
 typedef ISMRMRD::Image<complex_float_t> CFImage;
 typedef ISMRMRD::Image<complex_double_t> CDImage;
+
+#ifdef _MSC_VER
+#pragma warning( push )
+// disable warning about loss of precision. we convert to float in various places
+#pragma warning( disable : 4244 )
+#endif
 
 namespace sirf {
 
@@ -801,5 +807,9 @@ namespace sirf {
 		}
 	};
 }
+
+#ifdef _MSC_VER
+#pragma warning( pop )
+#endif
 
 #endif

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_x.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/gadgetron_x.h
@@ -37,41 +37,31 @@ limitations under the License.
 
 #include <ismrmrd/ismrmrd.h>
 #include <ismrmrd/dataset.h>
-#include <ismrmrd/meta.h>
-#include <ismrmrd/xml.h>
+
+#include "sirf/Gadgetron/cgadgetron_shared_ptr.h"
+#include "sirf/Gadgetron/gadgetron_data_containers.h"
 
 #include "sirf/common/JacobiCG.h"
 #include "sirf/iUtilities/LocalisedException.h"
 #include "sirf/Gadgetron/cgadgetron_shared_ptr.h"
-#include "sirf/Gadgetron/gadgetron_client.h"
 #include "sirf/Gadgetron/gadget_lib.h"
 #include "sirf/Gadgetron/ismrmrd_fftw.h"
 
 #define N_TRIALS 5
 
-/*!
-\ingroup Gadgetron Extensions
-\brief Shared pointer wrap-up for GadgetronClientConnector.
-
-*/
-
 namespace sirf {
 
+  class GadgetronClientConnector;
+
+	/*!
+	\ingroup Gadgetron Extensions
+	\brief Shared pointer wrap-up for GadgetronClientConnector.
+	*/
 	class GTConnector {
 	public:
-		GTConnector()
-		{
-			sptr_con_ = gadgetron::shared_ptr < GadgetronClientConnector >
-				(new GadgetronClientConnector);
-		}
-		GadgetronClientConnector& operator()()
-		{
-			return *sptr_con_.get();
-		}
-		gadgetron::shared_ptr<GadgetronClientConnector> sptr()
-		{
-			return sptr_con_;
-		}
+		GTConnector();
+		GadgetronClientConnector& operator()();
+		gadgetron::shared_ptr<GadgetronClientConnector> sptr();
 	private:
 		gadgetron::shared_ptr<GadgetronClientConnector> sptr_con_;
 	};

--- a/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/xgadgetron_utilities.h
+++ b/src/xGadgetron/cGadgetron/include/sirf/Gadgetron/xgadgetron_utilities.h
@@ -51,10 +51,14 @@ namespace sirf {
 		static std::string scratch_file_name()
 		{
 			static int calls = 0;
-			char buff[32];
+			char buff[128];
 			long long int ms = xGadgetronUtilities::milliseconds();
 			calls++;
+#ifdef _MSC_VER
+			sprintf_s(buff, 128, "tmp_%d_%lld.h5", calls, ms);
+#else
 			sprintf(buff, "tmp_%d_%lld.h5", calls, ms);
+#endif
 			return std::string(buff);
 		}
 		template<typename T>

